File: alpha/experiments/item-storage-protocol-standalone/src/exp/e4_statements.ts
// E4. Balance-forward statements — each month stands on the last, so disputes are
// bounded to one period. We run three billing periods with mixed activity (an add
// and a delete via manifest update, plus transfers), close each into a co-signed
// statement that chains to the previous by hash, then show that editing a
// historical figure breaks the chain at exactly that link and that a fabricated
// extra period cannot be inserted.

import type { World, ReceiptRecord } from "../world.ts";
import { Checker, type ExperimentResult } from "../types.ts";
import { makeItem } from "../item.ts";
import { makeReceipt, isAcknowledged } from "../receipt.ts";
import { buildStatement, verifyChain, type Statement } from "../statement.ts";
import { rentCents, postageCents } from "../pricing.ts";
import type { Json } from "../canonical.ts";

const PERIOD = 30;

export function run(world: World): ExperimentResult {
  const c = new Checker();
  const { customer, provider } = world;

  // A billable upload of a whole item: a bilateral receipt + a postage record.
  function billableUpload(cid: string, size: number): void {
    const core = {
      direction: "upload" as const, cid, byteStart: 0, byteEnd: size, bytes: size,
      runningTotal: size, day: world.clock.now(), receiverId: provider.id, senderId: customer.id,
    };
    const r = makeReceipt(core, provider.keypair, customer.keypair);
    if (!isAcknowledged(r)) throw new Error("upload receipt must be acknowledged");
    customer.ledger.append("receipt", world.clock.iso(), r as unknown as Json, [customer.signer()]);
    provider.ledger.append("receipt", world.clock.iso(), r as unknown as Json, [provider.signer()]);
    const rec: ReceiptRecord = {
      day: r.day, direction: r.direction, cid: r.cid, byteStart: 0, byteEnd: size,
      bytes: size, runningTotal: size,
    };
    world.receipts.push(rec);
  }

  // Close a period into a co-signed, chained statement (no audits yet — E5/E6).
  function close(startDay: number, endDay: number, openingRoot: string): Statement {
    const byteDays = world.byteDays(startDay, endDay);
    const postageBytes = world.receipts
      .filter((r) => r.day >= startDay && r.day < endDay)
      .reduce((s, r) => s + r.bytes, 0);
    const rc = rentCents(byteDays);
    const pc = postageCents(postageBytes);
    return world.commitStatement({
      periodStartDay: startDay, periodEndDay: endDay,
      openingRoot, closingRoot: world.manifest!.root, byteDays,
      rentCents: rc, postageBytes, postageCents: pc,
      auditCount: 0, auditBytes: 0, auditCents: 0, auditTier: "none",
      graceCents: 0, feesCents: 0, totalCents: rc + pc,
    });
  }

  // --- Period 0: steady state. Opening root = the E2 manifest; no change. ---
  const rootP0 = world.manifest!.root;
  world.clock.advanceDays(PERIOD); // day 0 -> 30
  const s0 = close(0, PERIOD, rootP0);

  // --- Period 1: Ada adds an item mid-period. ---
  const rootP1open = world.manifest!.root;
  world.clock.advanceDays(5); // day 30 -> 35
  const rng = world.rng("e4/new-item");
  const newBytes = Buffer.alloc(3000);
  for (let i = 0; i < newBytes.length; i++) newBytes[i] = rng.int(256);
  const added = makeItem("new-photo.jpg", newBytes);
  world.store.put(added);
  world.updateManifest([...world.items, added], "add new-photo.jpg");
  billableUpload(added.cid, added.size);
  world.clock.advanceDays(25); // day 35 -> 60
  const s1 = close(30, 60, rootP1open);

  // --- Period 2: Ada deletes an item mid-period. ---
  const rootP2open = world.manifest!.root;
  world.clock.advanceDays(10); // day 60 -> 70
  const removed = world.items.find((i) => i.label === "voice-note.ogg")!;
  world.store.drop(removed.cid);
  world.updateManifest(world.items.filter((i) => i.cid !== removed.cid), "delete voice-note.ogg");
  world.clock.advanceDays(20); // day 70 -> 90
  const s2 = close(60, 90, rootP2open);

  // The chain verifies end to end.
  c.ok("statement chain verifies from genesis", verifyChain(world.statements).ok);

  // Each statement's opening root equals the prior closing root (balance-forward).
  c.eq("period 1 opens where period 0 closed", s1.openingRoot, s0.closingRoot);
  c.eq("period 2 opens where period 1 closed", s2.openingRoot, s1.closingRoot);

  // Rent equals the independently recomputed byte-day integral.
  let rentMatches = true;
  for (const s of world.statements) {
    const recomputed = rentCents(world.byteDays(s.periodStartDay, s.periodEndDay));
    if (recomputed !== s.rentCents) rentMatches = false;
  }
  c.ok("rent equals the independently recomputed byte-day integral", rentMatches);

  // Adversarial: rewrite a figure inside statement 1. Chain must fail at link 1.
  const tampered: Statement[] = world.statements.map((s) => ({ ...s }));
  tampered[1] = { ...tampered[1], rentCents: tampered[1].rentCents + 100 };
  const tamperResult = verifyChain(tampered);
  c.ok("editing a historical figure fails chain verification", !tamperResult.ok);
  c.ok("chain failure is located at exactly the edited link",
    !tamperResult.ok && tamperResult.failedAt === 1);

  // Adversarial: insert a fabricated extra period between 1 and 2.
  const fabricated: Statement[] = [
    world.statements[0], world.statements[1],
    buildStatement({
      ...world.statements[2], period: 2, prevStatementHash: world.statements[1].hash,
      rentCents: 9999,
    }),
    world.statements[2],
  ];
  c.ok("a fabricated inserted period cannot pass chain verification",
    !verifyChain(fabricated).ok);

  return {
    id: "E4",
    title: "Balance-forward statements",
    plainSentence: "Last month was agreed, so this month only has to explain the difference.",
    assertions: c.results,
    tables: [
      {
        title: "Statement chain (cents)",
        headers: ["period", "byte-days", "rent", "postage", "total", "closing root"],
        rows: world.statements.map((s) => [
          s.period, s.byteDays, s.rentCents, s.postageCents, s.totalCents,
          s.closingRoot.slice(0, 12) + "…",
        ]),
      },
    ],
    notes: [],
  };
}

File: alpha/experiments/item-store/tests/e4_statements.rs
//! Phase 4 wiring test — ports E4 (balance-forward statements) from the oracle
//! (`src/exp/e4_statements.ts`): each period closes into a co-signed statement
//! (opening root, closing root, byte-day rent, postage) chained to the prior by
//! hash, so a dispute is bounded to one period. Adds the rollup/purge boundary:
//! once a period is settled, its granular receipts are purgeable while the chain
//! still verifies and rent stays recomputable.
//!
//! RED→GREEN gate: exercises item → manifest → rent-timeline → statement-chain
//! end to end, plus tamper/fabrication/purge.

use item_store::crypto::derive_keypair;
use item_store::identity::derive_id;
use item_store::item::Item;
use item_store::manifest::{build_manifest, ManifestLeaf};
use item_store::pricing::rent_cents;
use item_store::receipts::{make_bilateral_receipt, Direction, Receipt, ReceiptCore};
use item_store::statements::{
    build_statement, purge_receipts_settled_through, verify_chain, RentTimeline, Statement,
    StatementBody, GENESIS_STATEMENT,
};

const MASTER_SEED: &str = "item-store::e4::test-seed";

fn item_bytes(fill: u8, size: usize) -> Vec<u8> {
    (0..size).map(|i| fill ^ i.to_le_bytes()[0]).collect()
}

fn root_of(items: &[&Item], customer_id: &str, key: &item_store::crypto::Keypair) -> String {
    let leaves: Vec<ManifestLeaf> = items
        .iter()
        .map(|i| ManifestLeaf::new(i.cid(), i.size()))
        .collect();
    build_manifest(&leaves, customer_id, key).root().to_owned()
}

fn total_bytes(items: &[&Item]) -> u64 {
    items
        .iter()
        .map(|i| u64::try_from(i.size()).expect("size fits u64"))
        .sum()
}

/// Assemble and hash a period-closing statement from the rent timeline + the
/// period's receipts. `period = (num, start_day, end_day)`, `roots = (opening,
/// closing)`.
fn close(
    timeline: &RentTimeline,
    receipts: &[Receipt],
    period: (u64, u64, u64),
    roots: (&str, &str),
    prev_hash: &str,
) -> Statement {
    let (num, start, end) = period;
    let (opening, closing) = roots;
    let byte_days = timeline.byte_days(start, end);
    let postage_bytes: u64 = receipts
        .iter()
        .filter(|r| r.core().day >= start && r.core().day < end)
        .map(|r| u64::try_from(r.bytes()).expect("bytes fit u64"))
        .sum();
    let rent = rent_cents(byte_days);
    let postage = item_store::pricing::postage_cents(postage_bytes);
    build_statement(StatementBody {
        period: num,
        period_start_day: start,
        period_end_day: end,
        opening_root: opening.to_owned(),
        closing_root: closing.to_owned(),
        byte_days,
        rent_cents: rent,
        postage_bytes,
        postage_cents: postage,
        audit_count: 0,
        audit_bytes: 0,
        audit_cents: 0,
        audit_tier: "none".to_owned(),
        grace_cents: 0,
        fees_cents: 0,
        total_cents: rent + postage,
        prev_statement_hash: prev_hash.to_owned(),
    })
}

#[test]
#[allow(clippy::too_many_lines)] // ports E4's full three-period scenario + purge as one flow
fn e4_statements_chain_balance_forward_and_purge() {
    let customer = derive_keypair(MASTER_SEED, "customer");
    let customer_id = derive_id(&customer.verifying_key());

    // Base item set (period 0): A + B + C.
    let a = Item::from_bytes("will.pdf", item_bytes(0x11, 512));
    let b = Item::from_bytes("voice-note.ogg", item_bytes(0x22, 128));
    let c = Item::from_bytes("backup.tar", item_bytes(0x33, 8192));
    let root0 = root_of(&[&a, &b, &c], &customer_id, &customer);
    let total0 = total_bytes(&[&a, &b, &c]); // 8832

    let mut timeline = RentTimeline::new();
    timeline.set_bytes_at_rest(0, total0);

    let provider = derive_keypair(MASTER_SEED, "provider");
    let provider_id = derive_id(&provider.verifying_key());
    let mut receipts: Vec<Receipt> = Vec::new();

    // A billable upload of C on day 0 — postage inside period 0.
    let up0 = ReceiptCore::new(
        Direction::Upload,
        c.cid(),
        (0, c.size()),
        c.size(),
        0,
        &provider_id,
        &customer_id,
    );
    receipts.push(make_bilateral_receipt(up0, Some(&provider), &customer));

    // --- Period 0: steady state, day 0..30. ---
    let s0 = close(
        &timeline,
        &receipts,
        (0, 0, 30),
        (&root0, &root0),
        GENESIS_STATEMENT,
    );

    // --- Period 1: add item D mid-period (day 35), bill the upload. ---
    let d = Item::from_bytes("new-photo.jpg", item_bytes(0x44, 3000));
    let root1_open = root0.clone();
    let root1 = root_of(&[&a, &b, &c, &d], &customer_id, &customer);
    let total1 = total_bytes(&[&a, &b, &c, &d]); // 11832
    timeline.set_bytes_at_rest(35, total1);
    let up_core = ReceiptCore::new(
        Direction::Upload,
        d.cid(),
        (0, d.size()),
        d.size(),
        35,
        &provider_id,
        &customer_id,
    );
    receipts.push(make_bilateral_receipt(up_core, Some(&provider), &customer));
    let s1 = close(
        &timeline,
        &receipts,
        (1, 30, 60),
        (&root1_open, &root1),
        s0.hash(),
    );

    // --- Period 2: delete item B mid-period (day 70). ---
    let root2_open = root1.clone();
    let root2 = root_of(&[&a, &c, &d], &customer_id, &customer);
    let total2 = total_bytes(&[&a, &c, &d]); // 11704
    timeline.set_bytes_at_rest(70, total2);
    let s2 = close(
        &timeline,
        &receipts,
        (2, 60, 90),
        (&root2_open, &root2),
        s1.hash(),
    );

    let statements = vec![s0.clone(), s1.clone(), s2.clone()];

    // The chain verifies from genesis.
    assert!(
        verify_chain(&statements).is_ok(),
        "statement chain verifies from genesis"
    );

    // Balance-forward: each period opens where the prior closed.
    assert_eq!(
        s1.body().opening_root,
        s0.body().closing_root,
        "P1 opens where P0 closed"
    );
    assert_eq!(
        s2.body().opening_root,
        s1.body().closing_root,
        "P2 opens where P1 closed"
    );

    // Rent equals the independently recomputed byte-day integral.
    for s in &statements {
        let recomputed =
            rent_cents(timeline.byte_days(s.body().period_start_day, s.body().period_end_day));
        assert_eq!(
            recomputed,
            s.body().rent_cents,
            "rent == byte-day integral (period {})",
            s.body().period
        );
    }
    // Sanity on the arithmetic: P0 = 8832*30 byte-days.
    assert_eq!(s0.body().byte_days, total0 * 30);
    assert_eq!(s1.body().postage_bytes, 3000, "period 1 billed the upload");

    // Adversarial (a): edit a figure in statement 1 in place (keep its old hash)
    // → the chain fails at exactly link 1 (body no longer hashes to its hash).
    let mut edited = statements.clone();
    let mut body1 = edited[1].body().clone();
    body1.rent_cents += 100;
    edited[1] = Statement::from_parts(body1, edited[1].hash().to_owned());
    let result = verify_chain(&edited);
    assert!(
        !result.is_ok(),
        "editing a historical figure fails chain verification"
    );
    assert_eq!(
        result.failed_at(),
        Some(1),
        "failure located at exactly the edited link"
    );

    // Adversarial (b): a fabricated inserted period cannot pass.
    let mut fabricated = vec![s0.clone(), s1.clone()];
    fabricated.push(build_statement(StatementBody {
        period: 2,
        prev_statement_hash: s1.hash().to_owned(),
        rent_cents: 9999,
        ..s2.body().clone()
    }));
    fabricated.push(s2.clone());
    assert!(
        !verify_chain(&fabricated).is_ok(),
        "a fabricated inserted period cannot verify"
    );

    // Rollup + purge: period 0 is settled (closed + co-signed via s0). Its
    // granular receipts become purgeable — the signed statement carries the
    // rollup. Purge everything settled through day 30.
    let settled_postage = s0.body().postage_bytes; // the rollup carries it
    let purged = purge_receipts_settled_through(&mut receipts, 30);
    assert_eq!(
        purged, 1,
        "the day-0 receipt in the settled period is purged"
    );
    assert!(
        receipts.iter().all(|r| r.core().day >= 30),
        "no settled (day < 30) receipts remain",
    );
    // The signed statement still carries the period's rolled-up postage, and the
    // chain + rent survive the purge (rent comes from the timeline, not receipts).
    assert_eq!(
        s0.body().postage_bytes,
        settled_postage,
        "the co-signed rollup preserves the postage total"
    );
    assert!(
        verify_chain(&statements).is_ok(),
        "chain still verifies after purge"
    );
    assert_eq!(
        rent_cents(timeline.byte_days(0, 30)),
        s0.body().rent_cents,
        "rent for the purged period stays recomputable from the timeline",
    );
}
